### PR TITLE
Add a deprecatedPropType module to show deprecation warnings

### PR DESCRIPTION
**Pull request by janicduplessis**
To allow smoother API changes for users we often deprecate props and keep them around for a while before removing them. Right now it is all done manually, this adds a consistent way to show a warning when using a deprecated prop.

This also adds a deprecation warning of the website generated from the deprecatedPropType.

screen shot 2016-01-26 at 7 43 08 pm

It also changes places where we added the warnings manually to use deprecatedPropType instead.

### CODE REVIEW

**Comment by janicduplessis** [comment]
The only issue there is right now is that it generates 2 warnings, 1 for the JS component and 1 for the native one, I didn't really find a clean way to fix this since there is no way to make the difference between a native and normal component. One hacky way to do it would be to check if the component name starts with 'RCT' or add a prop on the propType object thats says it is a native component.

**Comment by satya164** [comment]
So awesome. Thanks @janicduplessis 

Checking `RCT` prefix won't work for many native components.

cc @mkonicek @nicklockwood

File: Libraries/Utilities/deprecatedPropType.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+'use strict';
+
+/**
+ * Adds a deprecation warning when the prop is used.
+ */
+function deprecatedPropType(
+  propType: ReactPropsCheckType,
+  explanation: string
+): ReactPropsCheckType {
```
Comment (janicduplessis) [inline Libraries/Utilities/deprecatedPropType.js]: Should we add this to eslint globals or is there a way to import these types?

**Comment by satya164** [comment]
I don't know where this type is defined. If it's in a file included via `.flowconfig`, we should define it as global. Otherwise `import type ReactPropsCheckType from './path'`

**Comment by janicduplessis** [comment]
@satya164 It's a type from react so it's builtin in flow. I guess the only way is to add it as a global in eslint.

**Comment by satya164** [comment]
@janicduplessis Yup!

**Comment by janicduplessis** [comment]
@satya164 I'll make a PR soon to add all flow global types to eslint so the bot stop complaining :)

**Comment by janicduplessis** [comment]
Ok I figured a clean way to check if it's a native module, I can just check if UIManager contains the component name.

**Comment by nicklockwood** [comment]
This looks great to me, but I think @vjeux is probably the best person to review it.

File: Libraries/Utilities/deprecatedPropType.js (hunk 1 above)
Comment (vjeux) [inline Libraries/Utilities/deprecatedPropType.js]: Add it to eslint global

File: website/server/docgenHelpers.js [hunk 2]
```diff
@@ -34,6 +35,43 @@ function stylePropTypeHandler(documentation, path) {
   });
 }
 
+function deprecatedPropTypeHandler(documentation, path) {
```
Comment (vjeux) [inline website/server/docgenHelpers.js]: So good

File: website/layout/AutodocsLayout.js [hunk 3]
```diff
@@ -108,6 +108,15 @@ var ComponentDoc = React.createClass({
             {renderType(prop.type)}
           </span>}
         </Header>
+        {prop.deprecationMessage && <div className="deprecated">
+          <div className="deprecatedTitle">
+            <img className="deprecatedIcon" src="/react-native/img/Warning.png" />
```
Comment (vjeux) [inline website/layout/AutodocsLayout.js]: Was the image already there? I don't see it in your commit

File: Libraries/Utilities/deprecatedPropType.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+const UIManager = require('UIManager');
+
+/**
+ * Adds a deprecation warning when the prop is used.
+ */
+function deprecatedPropType(
+  propType: ReactPropsCheckType,
+  explanation: string
+): ReactPropsCheckType {
+  return function validate(props, propName, componentName) {
```
Comment (vjeux) [inline Libraries/Utilities/deprecatedPropType.js]: I assume that this function is only going go be called when dev, but could you double check

**Comment by vjeux** [comment]
This is awesome, thanks for building this. Going to make our upgrade path much smoother :)

**Comment by vjeux** [comment]
@facebook-github-bot shipit

File: Libraries/Utilities/deprecatedPropType.js (hunk 1 above)
Comment (janicduplessis) [inline Libraries/Utilities/deprecatedPropType.js]: Fixed in #5585

Comment (vjeux) [inline Libraries/Utilities/deprecatedPropType.js]: What I thought but wanted to double check :)

File: website/layout/AutodocsLayout.js (hunk 3 above)
Comment (janicduplessis) [inline website/layout/AutodocsLayout.js]: Yes it's there in the first of the 2 commits at the bottom.

File: Libraries/Utilities/deprecatedPropType.js (hunk 4 above)
Comment (janicduplessis) [inline Libraries/Utilities/deprecatedPropType.js]: Double checked and this doesn't get called when __DEV__ is false :)

**Comment by janicduplessis** [comment]
Damn @vjeux in a 1300 lines css file we managed to both add new styles at the exact same place in the middle of the file :p

Rebased and fixed the conflict.

**Comment by vjeux** [comment]
Les grands esprits se rencontrent ;)

**Comment by vjeux** [comment]
@facebook-github-bot shipit

**Comment by janicduplessis** [comment]
Haha!

**Pull request closed without merging**